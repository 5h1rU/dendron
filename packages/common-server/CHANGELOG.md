# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.3.24](https://github.com/dendronhq/dendron/compare/v0.3.23...v0.3.24) (2020-07-20)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.21](https://github.com/dendronhq/dendron/compare/v0.3.20...v0.3.21) (2020-07-19)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.20](https://github.com/dendronhq/dendron/compare/v0.3.19...v0.3.20) (2020-07-18)

### Bug Fixes

- handle special characters in file names when importing using LocalFilePod ([03e42b1](https://github.com/dendronhq/dendron/commit/03e42b167ac9f073cd56f10c4e31b5cecf66dabf))

## [0.3.19](https://github.com/dendronhq/dendron/compare/v0.3.18...v0.3.19) (2020-07-18)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.18](https://github.com/dendronhq/dendron/compare/v0.3.17...v0.3.18) (2020-07-18)

### Features

- ability to import notes from local file system [docs](https://github.com/dendronhq/dendron-template/blob/master/vault/dendron.topic.pod.md) ([e44fd11](https://github.com/dendronhq/dendron/commit/e44fd1133bec10f22831f059c8d98cf4076dcdcc))

## [0.3.17](https://github.com/dendronhq/dendron/compare/v0.3.16...v0.3.17) (2020-07-17)

### Features

- more flexible file names for [journal notes](https://github.com/dendronhq/dendron-template/blob/master/vault/dendron.special-notes.md) ([9d9f10b](https://github.com/dendronhq/dendron/commit/9d9f10bd0873c201361c4625fc49c9f62ee82991))

## [0.3.16](https://github.com/dendronhq/dendron/compare/v0.3.15...v0.3.16) (2020-07-16)

### Features

- implement journal notes. see details here: https://github.com/dendronhq/dendron-template/blob/master/vault/dendron.feature.journals.md ([5e1236f](https://github.com/dendronhq/dendron/commit/5e1236fddbf1e0fddf4c27d1a40e9841cc99974f))

## [0.3.15](https://github.com/dendronhq/dendron/compare/v0.3.14...v0.3.15) (2020-07-14)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.11](https://github.com/dendronhq/dendron/compare/v0.3.10...v0.3.11) (2020-07-13)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.10](https://github.com/dendronhq/dendron/compare/v0.1.6...v0.3.10) (2020-07-13)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.9](https://github.com/dendronhq/dendron/compare/v0.3.8...v0.3.9) (2020-07-09)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.8](https://github.com/dendronhq/dendron/compare/v0.3.7...v0.3.8) (2020-07-09)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.7](https://github.com/dendronhq/dendron/compare/v0.3.6...v0.3.7) (2020-07-08)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.6](https://github.com/dendronhq/dendron/compare/v0.3.5...v0.3.6) (2020-07-07)

**Note:** Version bump only for package @dendronhq/common-server

# [0.3.0](https://github.com/dendronhq/dendron/compare/v0.2.20...v0.3.0) (2020-07-05)

**Note:** Version bump only for package @dendronhq/common-server
