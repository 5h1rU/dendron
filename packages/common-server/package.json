{
  "name": "@dendronhq/common-server",
  "version": "0.1.6",
  "description": "common-server",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/dendronhq/dendron.git",
    "directory": "packages/common-server"
  },
  "author": {
    "name": "Kevin Lin",
    "email": "kevin@thence.io",
    "url": "https://kevinslin.com",
    "organization": true
  },
  "keywords": [
    ""
  ],
  "files": [
    "lib"
  ],
  "main": "lib/index",
  "types": "lib/index",
  "scripts": {
    "clean": "rimraf lib && rimraf coverage",
    "format": "",
    "lint": "",
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean && npm run format && npm run lint && echo Using TypeScript && tsc --version",
    "build": "npm run compile",
    "compile": "tsc -p tsconfig.build.json",
    "test": "jest",
    "coverage": "jest --coverage",
    "watch": "npm run compile -- --watch",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@dendronhq/common-all": "^0.1.6",
    "gray-matter": "^4.0.2",
    "lodash": "^4.17.15",
    "minimatch": "^3.0.4",
    "pino": "^6.3.2",
    "yamljs": "^0.3.0"
  },
  "devDependencies": {
    "@types/pino": "^6.3.0",
    "@types/jest": "^23.3.3",
    "@types/lodash": "^4.14.154",
    "@types/minimatch": "^3.0.3",
    "@types/node": "^10.11.4",
    "@types/yamljs": "^0.2.31",
    "coveralls": "^3.0.2",
    "jest": "^23.6.0",
    "prettier": "^1.14.3",
    "rimraf": "^2.6.2",
    "ts-jest": "^23.10.3",
    "ts-node": "^7.0.1",
    "tslint": "^5.11.0",
    "tslint-config-prettier": "^1.15.0",
    "typescript": "^3.1.1"
  },
  "engines": {
    "node": ">=10.0.0"
  },
  "jest": {
    "preset": "ts-jest"
  },
  "gitHead": "403eb012f8b34bc115dd7dd86f80c76b39329f3b"
}
