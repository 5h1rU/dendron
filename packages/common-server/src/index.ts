export * from "./files";
export * from "./logger";
export { testUtils, FileTestUtils } from "./testUtils";
