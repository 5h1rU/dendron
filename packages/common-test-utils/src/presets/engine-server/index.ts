import { ENGINE_DELETE_PRESETS } from "./delete";
import ENGINE_MULTI_TEST_PRESET from "./engine-multi";
import ENGINE_SINGLE_TEST_PRESET from "./engine-single";
import { ENGINE_GET_NOTE_BY_PATH_PRESETS } from "./getByPath";
import { ENGINE_INIT_PRESETS } from "./init";
import NOTE_REF from "./note-refs";
import { ENGINE_QUERY_PRESETS } from "./query";
import { ENGINE_RENAME_PRESETS } from "./rename";
import { ENGINE_UPDATE_PRESETS } from "./update";
import { ENGINE_WRITE_PRESETS } from "./write";

export const ENGINE_SERVER = {
  NOTE_REF,
  ENGINE_MULTI_TEST_PRESET,
  ENGINE_SINGLE_TEST_PRESET,
  ENGINE_WRITE_PRESETS,
  ENGINE_INIT_PRESETS,
  ENGINE_UPDATE_PRESETS,
  ENGINE_DELETE_PRESETS,
  ENGINE_GET_NOTE_BY_PATH_PRESETS,
  ENGINE_RENAME_PRESETS,
  ENGINE_QUERY_PRESETS,
};

export const ENGINE_PRESETS = [
  // {name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS},
  // {name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS},
  // {name: "getByPath", presets: ENGINE_SERVER.ENGINE_GET_NOTE_BY_PATH_PRESETS},
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
  // {name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS},
  // {name: "update", presets: ENGINE_SERVER.ENGINE_UPDATE_PRESETS},
  // {name: "write", presets: ENGINE_SERVER.ENGINE_WRITE_PRESETS},
];
