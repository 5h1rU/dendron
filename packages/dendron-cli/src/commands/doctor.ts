import {
  DendronASTDest,
  MDUtilsV4,
  RemarkChangeEntry,
  RemarkUtils,
} from "@dendronhq/engine-server";
import _ from "lodash";
import yargs from "yargs";
import { CLICommand } from "./base";
import { CommandOptsV3 } from "./soil";
import { setupEngine } from "./utils";
import throttle from "@jcoreio/async-throttle";

type CommandCLIOpts = {
  wsRoot: string;
  actions: DoctorActions[];
  enginePort?: number;
  query?: string;
  limit?: number;
};

type CommandOpts = CommandOptsV3 & CommandCLIOpts;
type CommandOutput = void;

export enum DoctorActions {
  H1_TO_TITLE = "h1ToTitle",
  HI_TO_H2 = "h1ToH2",
}

export class DoctorCLICommand extends CLICommand<CommandOpts, CommandOutput> {
  constructor() {
    super({ name: "doctor", desc: "doctor helps you fix your notes" });
  }

  buildArgs(args: yargs.Argv) {
    super.buildArgs(args);
    args.option("actions", {
      describe: "what actions the doctor should take",
      requiresArg: true,
      type: "array",
      choices: Object.values(DoctorActions),
    });
    args.option("enginePort", {
      describe: "port that engine is running on",
    });
    args.option("query", {
      describe: "run doctor over a query",
      type: "string",
    });
    args.option("limit", {
      describe: "limit num changes",
      type: "number",
    });
  }

  async enrichArgs(args: CommandCLIOpts): Promise<CommandOpts> {
    const engineArgs = await setupEngine(args);
    return { ...args, ...engineArgs };
  }

  async execute(opts: CommandOpts) {
    const { actions, engine, query, limit } = _.defaults(opts, {
      limit: 99999,
    });
    const proc = MDUtilsV4.procFull({
      dest: DendronASTDest.MD_DENDRON,
      engine,
      mathOpts: { katex: true },
    });
    let notes = query
      ? engine.queryNotesSync({ qs: query }).data
      : _.values(engine.notes);
    this.L.info({ msg: "prep doctor", numResults: notes.length });
    let numChanges = 0;
    let notesSeeked = 0;
    await _.reduce<any, Promise<any>>(
      actions,
      async (acc, action) => {
        await acc;
        switch (action) {
          case DoctorActions.H1_TO_TITLE: {
            return Promise.all(
              notes.map(async (note) => {
                this.L.info({ msg: `processing ${note.fname}` });
                let changes: RemarkChangeEntry[] = [];
                const newBody = await proc()
                  .use(RemarkUtils.h1ToTitle(note, changes))
                  .process(note.body);
                note.body = newBody.toString();
                this.L.info({ msg: `changes ${note.fname}`, changes });
                return;
              })
            );
          }
          case DoctorActions.HI_TO_H2: {
            const engineWrite = throttle(
              _.bind(engine.writeNote, engine),
              5000,
              { leading: true }
            );
            return _.reduce<any, Promise<any>>(
              notes,
              async (accInner, note) => {
                await accInner;
                if (numChanges >= limit) {
                  return;
                }
                this.L.debug({ msg: `processing for h1_to_h2 ${note.fname}` });
                let changes: RemarkChangeEntry[] = [];
                const newBody = await proc()
                  .use(RemarkUtils.h1ToH2(changes))
                  .process(note.body);
                note.body = newBody.toString();
                if (!_.isEmpty(changes)) {
                  await engineWrite(note);
                  this.L.info({ msg: `changes ${note.fname}`, changes });
                  numChanges += 1;
                  return;
                } else {
                  return;
                }
              },
              Promise.resolve()
            );
          }
        }
        return;
      },
      Promise.resolve()
    );
    this.L.info({ msg: "doctor done", numChanges });
    return;
  }
}

// export class DoctorCommand extends SoilCommandV3<
//   CommandCLIOpts,
//   CommandOpts,
//   CommandOutput
// > {

//   static buildCmd(yargs: yargs.Argv): yargs.Argv {
//     const _cmd = new DoctorCommand();
//     return yargs.command(
//       "doctor",
//       "run doctor command",
//       _cmd.buildArgs,
//       _cmd.eval
//     );
//   }

//   buildArgs(args: yargs.Argv) {
//     super.buildArgs(args);
//     args.option("actions", {
//       describe: "what actions the doctor should take",
//       requiresArg: true,
//       type: "array",
//       choices: ["h1ToTitle", "otherAction"]
//     });
//     args.option("enginePort", {
//       describe: "port that engine is running on"
//     });
//   }

//   async enrichArgs(args: CommandCLIOpts): Promise<CommandOpts> {
//     console.log("setupengine pre");
//     const engineArgs = await setupEngine(args);
//     console.log("setupengine post");
//     return { ...args, ...engineArgs };
//   }

//   async execute(opts: CommandOpts) {
//     console.log("engine init");
//     const { actions, engine } = opts;
//     const proc = MDUtilsV4.procFull({
//       dest: DendronASTDest.MD_DENDRON,
//       engine,
//       mathOpts: { katex: true }
//     });
//     await _.reduce<any, Promise<any>>(
//       actions,
//       async (acc, action) => {
//         await acc;
//         if (action === DoctorActions.H1_TO_TITLE) {
//           return Promise.all(
//             _.values(engine.notes).map(async note => {
//               const newBody = await proc()
//                 .use(RemarkUtils.h1ToTitle(note))
//                 .process(note.body);
//               note.body = newBody.toString();
//               await engine.writeNote(note, { updateExisting: true });
//               return;
//             })
//           );
//         }
//         return;
//       },
//       Promise.resolve()
//     );
//     return;
//   }
// }
