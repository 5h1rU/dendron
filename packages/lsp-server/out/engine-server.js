"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DendronEngineServer = void 0;
const engine_server_1 = require("@dendronhq/engine-server");
class DendronEngineServer {
    constructor(opts) {
        this.opts = opts;
    }
    static getOrCreate(opts) {
        if (!DendronEngineServer._instance) {
            DendronEngineServer._instance = new DendronEngineServer(opts);
        }
        return DendronEngineServer._instance;
    }
    init() {
        const root = this.opts.roots[0];
        if (!root) {
            return;
        }
        const engine = engine_server_1.DendronEngine.getOrCreateEngine({
            root: root.uri,
            forceNew: true,
        });
        this._engine = engine;
        this._engine.init();
    }
}
exports.DendronEngineServer = DendronEngineServer;
//# sourceMappingURL=engine-server.js.map