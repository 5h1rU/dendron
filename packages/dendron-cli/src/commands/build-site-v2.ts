import { DEngineClientV2 } from "@dendronhq/common-all";
import { goUpTo } from "@dendronhq/common-server";
import { SiteUtils } from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import yargs from "yargs";
import { CLICommand } from "./base";
import { setupEngine, setupEngineArgs } from "./utils";

type CommandCLIOpts = {
  wsRoot: string;
  port?: number;
  engine?: DEngineClientV2;
  cwd?: string;
  servePort?: number;
  enginePort?: number;
  serve: boolean;
  stage: "dev" | "prod";
  output?: string;
};
type CommandOpts = CommandCLIOpts & { engine: DEngineClientV2 };
type CommandOutput = {};

export { CommandOpts as BuildSiteV2CLICommandOpts };
export class BuildSiteV2CLICommand extends CLICommand<
  CommandOpts,
  CommandOutput
> {
  constructor() {
    super({
      name: "buildSiteV2",
      desc: "build notes for publication using 11ty",
    });
  }

  buildArgs(args: yargs.Argv) {
    super.buildArgs(args);
    setupEngineArgs(args);
    args.option("serve", {
      describe: "serve over local http server",
      default: false,
      type: "boolean",
    });
    args.option("stage", {
      describe: "serve over local http server",
      default: "dev",
      choices: ["dev", "prod"],
    });
    args.option("servePort", {
      describe: "port to serve over",
      default: "8080",
    });
    args.option("output", {
      describe: "if set, override output from config.yml",
      type: "string",
    });
  }

  async enrichArgs(args: CommandCLIOpts): Promise<CommandOpts> {
    const engineArgs = await setupEngine(args);
    this.L.info({ msg: `connecting to engine on port: ${engineArgs.port}` });
    // add site specific notes
    if (args.enginePort) {
      const siteNotes = SiteUtils.addSiteOnlyNotes({
        engine: engineArgs.engine,
      });
      _.forEach(siteNotes, (ent) => {
        engineArgs.engine.notes[ent.id] = ent;
      });
    }
    return { ...args, ...engineArgs };
  }

  async execute(opts: CommandOpts) {
    let nmPath = goUpTo(__dirname, "node_modules");
    let { wsRoot, port, stage, servePort, output } = _.defaults(opts, {});
    let cwd = opts.cwd;
    if (!cwd) {
      cwd = path.join(nmPath, "node_modules", "@dendronhq", "dendron-11ty");
      // fix for /home/runner/work/dendron-site/dendron-site/node_modules/@dendronhq/dendron-cli/node_modules/@dendronhq/dendron-11ty'
      if (!fs.existsSync(cwd)) {
        nmPath = goUpTo(path.join(nmPath, ".."), "node_modules");
        cwd = path.join(nmPath, "node_modules", "@dendronhq", "dendron-11ty");
      }
    }
    process.env["ENGINE_PORT"] = _.toString(port);
    process.env["WS_ROOT"] = wsRoot;
    process.env["STAGE"] = stage;
    process.env["ELEV_PORT"] = _.toString(servePort);
    if (output) {
      process.env["OUTPUT"] = output;
    }
    const { compile } = require("@dendronhq/dendron-11ty");
    await compile({ cwd }, { serve: opts.serve, port: servePort });
    if (!opts.serve) {
      // hack, give postBuild a chance to complete
      setTimeout(() => {
        process.exit();
      }, 5000);
    }
    return {};
  }
}
