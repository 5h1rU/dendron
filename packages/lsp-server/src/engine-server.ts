import { WorkspaceFolder } from "vscode-languageserver";
import { DendronEngine } from "@dendronhq/engine-server";

type DendronEngineServerOpts = {
  roots: WorkspaceFolder[];
};

export class DendronEngineServer {
  protected _engine?: DendronEngine;

  constructor(public opts: DendronEngineServerOpts) {}

  static _instance: DendronEngineServer | null;

  static getOrCreate(opts: DendronEngineServerOpts) {
    if (!DendronEngineServer._instance) {
      DendronEngineServer._instance = new DendronEngineServer(opts);
    }
    return DendronEngineServer._instance;
  }

  init() {
    const root = this.opts.roots[0];
    if (!root) {
      return;
    }
    const engine = DendronEngine.getOrCreateEngine({
      root: root.uri,
      forceNew: true,
    }) as DendronEngine;
    this._engine = engine;
    this._engine.init();
  }
}
