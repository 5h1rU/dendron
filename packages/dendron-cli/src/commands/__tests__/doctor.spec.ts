import { file2Note, note2File, tmpDir } from "@dendronhq/common-server";
import {
  ENGINE_HOOKS,
  NoteTestUtilsV4,
  runEngineTestV4,
} from "@dendronhq/common-test-utils";
import { createEngine } from "@dendronhq/engine-server";
import { ConfigUtils } from "@dendronhq/engine-test-utils";
import path from "path";
import { BuildSiteCommandV2 } from "../build-site-v2";
import { DoctorActions, DoctorCommand } from "../doctor";

describe("basic", () => {
  let siteRootDir: string;
  beforeEach(async () => {
    siteRootDir = tmpDir().name;
  });

  test("basic", async () => {
    await runEngineTestV4(
      async ({ engine, wsRoot, vaults }) => {
        const vault = vaults[0];
        const cmd = new DoctorCommand();
        await cmd.execute({
          wsRoot,
          engine,
          actions: [DoctorActions.H1_TO_TITLE],
        });
        const fpath = path.join(wsRoot, vault.fsPath, "foo.md");
        const note = file2Note(fpath, vault);
        expect(note).toMatchSnapshot();
      },
      {
        createEngine,
        expect,
        preSetupHook: async ({ wsRoot, vaults }) => {
          //const txt = `# Hello Heading\nHello Content`;
          await NoteTestUtilsV4.createNote({
            wsRoot,
            vault: vaults[0],
            fname: "foo",
            body: [`# Foo Header`, `## Foo Content`].join("\n"),
          });
          await NoteTestUtilsV4.createNote({
            wsRoot,
            vault: vaults[0],
            fname: "bar",
            body: [`# Bar Header`, `## Bar Content`].join("\n"),
          });
        },
      }
    );
  });
});
