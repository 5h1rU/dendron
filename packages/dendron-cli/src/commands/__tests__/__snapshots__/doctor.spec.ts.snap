// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic basic 1`] = `
Object {
  "body": "## Foo Content
",
  "children": Array [],
  "created": "1",
  "custom": Object {},
  "data": Object {},
  "desc": "",
  "fname": "foo",
  "id": "foo",
  "links": Array [],
  "parent": null,
  "title": "Foo Header",
  "type": "note",
  "updated": "1",
  "vault": Object {
    "fsPath": "vault1",
  },
}
`;
